reframe/frontend/executors/__init__.py

```diff
@@ -168,14 +180,18 @@ def runall(self, checks, system):
             self._printer.timestamp('Started on', 'short double line')
             self._printer.info()
             self._runall(checks, system)
+            if self._max_retries > 0:
+                self._retry(checks, system)
+
         finally:
-            # Always update statistics and print the summary line
-            self._stats = self._policy.getstats()
-            num_failures = self._stats.num_failures()
+            # Print the summary line
+            num_failures_last_retry = self._stats.num_failures(try_num=-1)
             self._printer.status(
-                'FAILED' if num_failures else 'PASSED',
-                'Ran %d test case(s) from %d check(s) (%d failure(s))' %
-                (self._stats.num_cases(), len(checks), num_failures),
+                'FAILED' if num_failures_last_retry else 'PASSED',
+                'Ran %d test case(s) from %d check(s) (%d failure(s) after '
+                '%d retries; %d test case(s) passed in retries)' %
+                (self._stats.num_cases(), len(checks), num_failures_last_retry,
+                 self._try_num, self._stats.num_success_all_retries()),
```

Comment:
I would keep the old message as it was and add a `retry_report()` method in the statistics. This report would print "No regression test has been retried." if none was retried otherwise it should print sth like the following:

```
RETRY SUMMARY
- Test blah_blah_test has been retried 1 time(s) and passed.
- Test blouh_blouh_test has been retried 2 time(s) and failed.
```

Reply:
I organised the output as follows (the points in parenthesis are not printed if there are no retries):

SUMMARY OF FAILURES
1. global stats on failures
(2. global stats on retries)
3. List of failed test
(4. List of tests passed in retry x (only printed if any test passed in retry x))
5. FAILURE INFO for each failed test

Reply:
I would like to keep the reports separate as in my original proposal:

```
[The output of the runs as before + the additional runs]
SUMMARY OF RETRIES
    - Test foo_check on daint:gpu using PrgEnv-gnu was retried 1 time(s) and passed.
    - Test foo_check on daint:gpu using PrgEnv-cray was retried 2 time(s) and failed.
SUMMARY OF FAILURES
[as before]
```
The failure summary should only be printed if there are failures (as of now). The string `foo_check on daint:gpu using PrgEnv-gnu` may be easily retrieved by calling the `info()` method of the check.

Reply:
OK. One thing though: at no place appears the number of failed test cases. How about, I add it 
in this line
```
[  FAILED  ] Ran 2 test case(s) from 2 check(s)
```
e.g. something like
```
[  FAILED  ] Ran 2 test case(s) from 2 check(s) with 1 failure(s)
```
?

Reply:
This status line should be unchanged, exactly as before. It contained already the failure count information. What I am suggesting here is to put the retry information in a separate section.

Reply:
I added the separate "SUMMARY OF RETRIES" section following your example. However, if there are no retries at all, then I would rather completely omit the SUMMARY OF RETRIES" section instead of printing "No regression test has been retried.". Do you agree?